model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanKeycloakTransaction.java

```diff
@@ -122,10 +122,12 @@ public <K, V> void replace(Cache<K, V> cache, K key, V value, long lifespan, Tim
 
         Object taskKey = getTaskKey(cache, key);
         CacheTask current = tasks.get(taskKey);
-        if (current != null) {
+        if (current != null && current != TOMBSTONE && current.getOperation() != Operation.REMOVE) {
```

Comment:
@ahus1 I just realized we are changing the semantics of this method with this change. Previously when there was REMOVE and then REPLACE operation called on an existing entry it resulted with REMOVE of the entry. While now the result is REPLACE and DELETE is ignored. 

Is this change intended? Should we change the behavior in a patch release?

Reply:
@mhajas - yes, this was intended. Once I added the 

```
throw new IllegalStateException("Can't replace entry: task " + current + " in progress for session");
```

some of the concurrency-tests started to fail. Looking at the (isolated) logic, a replace following a previous delete should do a replace, and not keep the original delete. I know this is the logic present in all previous releases, but it is really weird. Do we want to dive deeper on this and see which tests fail when an update is silently ignored when there was a previous remove?

Reply:
I looked deeper into this, and it seems that `RootAuthenticationSessionAdapter` will call update even after it has called delete. Thank you for asking your questions here. 

I'm closing this backport, and will open a bug issue against main.